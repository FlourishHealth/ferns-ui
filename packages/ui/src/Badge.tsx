import React from "react";
import {Text, View} from "react-native";

import {BadgeProps, SurfaceTheme, TextTheme} from "./Common";
import {Icon} from "./Icon";
import {useTheme} from "./Theme";

export const Badge = ({
  value,
  iconName,
  status = "info",
  secondary = false,
  variant,
  maxValue = 100,
  customBackgroundColor,
  customTextColor,
  customBorderColor,
  customIconColor,
  customIconName,
  testID,
}: BadgeProps): React.ReactElement => {
  const {theme} = useTheme();
  const isIconOnly = variant === "iconOnly";

  let badgeColor: keyof TextTheme = "inverted";

  // TODO: Move to theme
  const secondaryBorderColors = {
    error: "#F39E9E",
    warning: "#FCC58F",
    info: "#8FC1D2",
    success: "#7FD898",
    neutral: "#AAAAAA",
    custom: "#AAAAAA",
  };

  if (secondary) {
    if (status === "error") {
      badgeColor = "error";
    } else if (status === "warning") {
      badgeColor = "warning";
    } else if (status === "info") {
      badgeColor = "secondaryDark";
    } else if (status === "success") {
      badgeColor = "success";
    } else if (status === "neutral") {
      badgeColor = "primary";
    }
  }

  let badgeBgColor: keyof SurfaceTheme = "neutralDark";

  if (status === "error") {
    badgeBgColor = secondary ? "errorLight" : "error";
  } else if (status === "warning") {
    badgeBgColor = secondary ? "warningLight" : "warning";
  } else if (status === "info") {
    badgeBgColor = secondary ? "secondaryLight" : "secondaryDark";
  } else if (status === "success") {
    badgeBgColor = secondary ? "successLight" : "success";
  } else if (status === "neutral") {
    badgeBgColor = secondary ? "neutralLight" : "neutralDark";
  }

  const backgroundColor = status === "custom" ? customBackgroundColor : theme.surface[badgeBgColor];
  const borderColor = status === "custom" ? customBorderColor : secondaryBorderColors[status];
  const textColor = status === "custom" ? customTextColor : theme.text[badgeColor];
  const iconColor = status === "custom" ? customIconColor : badgeColor;

  let badgeBorderRadius = theme.radius.default;
  if (isIconOnly) {
    badgeBorderRadius = theme.radius.full;
  } else if (variant === "numberOnly") {
    badgeBorderRadius = theme.radius.rounded;
  }

  let badgeValue = value;

  if (variant === "numberOnly") {
    if (!isNaN(Number(value)) && maxValue) {
      const numberValue = Number(value);
      if (numberValue > maxValue) {
        badgeValue = `${maxValue}+`;
      } else {
        badgeValue = numberValue;
      }
    } else {
      console.warn("Warning: Badge value is not a number");
    }
  }

  return (
    <View
      style={{
        alignItems: "flex-start",
      }}
    >
      <View
        style={[
          {
            justifyContent: "center",
            alignItems: "center",
            paddingVertical: variant === "iconOnly" ? 1 : theme.spacing.xs,
            paddingHorizontal: variant === "iconOnly" ? theme.spacing.xs : theme.spacing.sm,
            flexDirection: "row",
            borderRadius: badgeBorderRadius,
            backgroundColor,
            height: variant === "iconOnly" ? 16 : "auto",
            width: variant === "iconOnly" ? 16 : "auto",
          },
          isIconOnly && {height: 16, width: 16},
          secondary && {borderWidth: 1, borderColor},
        ]}
        testID={testID}
      >
        {Boolean(variant !== "numberOnly" && iconName) && (
          <View style={{marginRight: variant === "iconOnly" ? 0 : theme.spacing.sm}} testID="icon">
            <Icon color={iconColor} iconName={customIconName ?? iconName!} size="xs" />
          </View>
        )}
        {Boolean(variant !== "iconOnly") && (
          <Text
            style={{
              color: textColor,
              fontSize: 10,
              fontWeight: "700",
              fontFamily: "text",
            }}
          >
            {badgeValue}
          </Text>
        )}
      </View>
    </View>
  );
};
