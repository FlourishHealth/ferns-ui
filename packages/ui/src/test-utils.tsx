import {render} from "@testing-library/react-native";
import React from "react";
import {ThemeProvider} from "./Theme";

export const renderWithTheme = (ui: React.ReactElement) => {
  return render(<ThemeProvider>{ui}</ThemeProvider>);
};

export const createCommonMocks = () => ({
  onChange: jest.fn(),
  onFocus: jest.fn(),
  onBlur: jest.fn(),
  onEnter: jest.fn(),
  onSubmitEditing: jest.fn(),
  onIconClick: jest.fn(),
});

export const setupComponentTest = () => {
  jest.useFakeTimers();
  return createCommonMocks();
};

export const teardownComponentTest = () => {
  jest.useRealTimers();
  jest.clearAllMocks();
};
