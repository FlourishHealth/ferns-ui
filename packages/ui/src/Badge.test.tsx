import React from "react";

import {Badge} from "./Badge";
import {renderWithTheme} from "./test-utils";

describe("Badge", () => {
  const defaultProps = {
    value: "Test",
  };

  it("renders correctly with default props", () => {
    const {toJSON} = renderWithTheme(<Badge {...defaultProps} />);
    expect(toJSON()).toMatchSnapshot();
  });

  it("renders text value correctly", () => {
    const {getByText} = renderWithTheme(<Badge {...defaultProps} />);
    expect(getByText("Test")).toBeTruthy();
  });

  it("renders number value correctly", () => {
    const {getByText} = renderWithTheme(<Badge value={42} />);
    expect(getByText("42")).toBeTruthy();
  });

  it("truncates large numbers with maxValue", () => {
    const {getByText} = renderWithTheme(<Badge maxValue={100} value={150} variant="numberOnly" />);
    expect(getByText("100+")).toBeTruthy();
  });

  it("does not truncate numbers below maxValue", () => {
    const {getByText} = renderWithTheme(<Badge maxValue={100} value={50} variant="numberOnly" />);
    expect(getByText("50")).toBeTruthy();
  });

  it("applies correct status colors", () => {
    const statuses = ["error", "warning", "info", "success", "neutral"] as const;

    statuses.forEach((status) => {
      let {getByTestId} = renderWithTheme(
        <Badge {...defaultProps} status={status} testID="badge" />
      );

      // Test primary variant
      const badge = getByTestId("badge");
      expect(badge).toHaveStyle({backgroundColor: expect.any(String)});

      // Test secondary variant
      ({getByTestId} = renderWithTheme(
        <Badge {...defaultProps} secondary status={status} testID="badge-secondary" />
      ));
      const secondaryBadge = getByTestId("badge-secondary");
      expect(secondaryBadge).toHaveStyle({backgroundColor: expect.any(String)});
    });
  });

  it("renders icon when iconName is provided", () => {
    const {getByTestId} = renderWithTheme(
      <Badge {...defaultProps} iconName="check" testID="badge-with-icon" />
    );
    expect(getByTestId("icon")).toBeTruthy();
  });

  it("renders icon only when variant is iconOnly", () => {
    const {getByTestId, queryByText} = renderWithTheme(
      <Badge {...defaultProps} iconName="check" testID="icon-only-badge" variant="iconOnly" />
    );

    expect(getByTestId("icon")).toBeTruthy();
    expect(queryByText("Test")).toBeNull();
  });

  it("applies custom colors when status is custom", () => {
    const customColors = {
      customBackgroundColor: "#123456",
      customTextColor: "#ffffff",
      customBorderColor: "#654321",
      customIconColor: "#ffcc00",
    };

    const {getByTestId} = renderWithTheme(
      <Badge
        {...defaultProps}
        iconName="star"
        secondary
        status="custom"
        testID="custom-badge"
        {...(customColors as any)}
      />
    );

    const badge = getByTestId("custom-badge");
    expect(badge).toHaveStyle({
      backgroundColor: customColors.customBackgroundColor,
      borderColor: customColors.customBorderColor,
    });
  });

  it("applies correct border radius based on variant", () => {
    // Default variant
    let {getByTestId} = renderWithTheme(<Badge {...defaultProps} testID="default-badge" />);
    expect(getByTestId("default-badge")).toHaveStyle({borderRadius: expect.any(Number)});

    // Icon only variant
    ({getByTestId} = renderWithTheme(
      <Badge {...defaultProps} iconName="check" testID="icon-only-badge" variant="iconOnly" />
    ));
    expect(getByTestId("icon-only-badge")).toHaveStyle({borderRadius: expect.any(Number)});

    // Number only variant
    ({getByTestId} = renderWithTheme(
      <Badge {...defaultProps} testID="number-only-badge" variant="numberOnly" />
    ));
    expect(getByTestId("number-only-badge")).toHaveStyle({borderRadius: expect.any(Number)});
  });
});
