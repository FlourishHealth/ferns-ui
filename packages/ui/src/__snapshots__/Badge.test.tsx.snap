// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Badge renders correctly with default props 1`] = `
<View
  style={
    {
      "alignItems": "flex-start",
    }
  }
>
  <View
    style={
      [
        {
          "alignItems": "center",
          "backgroundColor": "#2B6072",
          "borderRadius": 4,
          "flexDirection": "row",
          "height": "auto",
          "justifyContent": "center",
          "paddingHorizontal": 8,
          "paddingVertical": 4,
          "width": "auto",
        },
        false,
        false,
      ]
    }
  >
    <Text
      style={
        {
          "color": "#FFFFFF",
          "fontFamily": "text",
          "fontSize": 10,
          "fontWeight": "700",
        }
      }
    >
      Test
    </Text>
  </View>
</View>
`;
